File: src/main/java/fromjavatokotlin/service/RobotService.java
package fromjavatokotlin.service;

import fromjavatokotlin.model.Robot;
import fromjavatokotlin.repository.RobotMongoRepo;
import lombok.RequiredArgsConstructor;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.data.mongodb.core.query.Update;
import org.springframework.stereotype.Service;

import javax.annotation.Nonnull;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static fromjavatokotlin.constant.DocumentConstants.ID_FIELD;
import static fromjavatokotlin.model.Robot.ABILITIES_FIELD;
import static org.springframework.data.mongodb.core.query.Criteria.where;

@Service
@RequiredArgsConstructor
public class RobotService {

  private final RobotMongoRepo robotMongoRepo;

  private final MongoTemplate mongoTemplate;

  public List<Robot> findAll() {
    return robotMongoRepo.findAll();
  }

  public List<Robot> findOnlyAvailable() {
    return robotMongoRepo.findOnlyAvailable();
  }

  public void updateAvailability(@Nonnull String id, boolean availability) {
    checkNotNull(id);
    mongoTemplate.updateFirst(
        new Query(where(ID_FIELD).is(id)),
        Update.update(Robot.AVAILABILITY_FIELD, availability),
        Robot.class
    );
  }

  public void addAbility(/*@Nonnull*/ String id, /*@Nonnull*/ Robot.Ability ability) {
    checkNotNull(id);
    checkNotNull(ability);
    mongoTemplate.updateFirst(
        new Query(where(ID_FIELD).is(id)),
        new Update().push(ABILITIES_FIELD, ability),
        Robot.class
    );
  }
}

File: src/main/kotlin/fromjavatokotlin/service/RobotService.kt
package fromjavatokotlin.service

import fromjavatokotlin.constant.DocumentConstants
import fromjavatokotlin.model.Robot
import fromjavatokotlin.model.Robot.Ability
import fromjavatokotlin.repository.RobotMongoRepo
import fromjavatokotlin.util.updateById
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import org.springframework.data.mongodb.core.query.Update
import org.springframework.data.mongodb.core.query.isEqualTo
import org.springframework.data.mongodb.core.update
import org.springframework.stereotype.Service

@Service
class RobotService(
    private val robotMongoRepo: RobotMongoRepo,
    private val mongoTemplate: MongoTemplate
) {
    fun findAll(): List<Robot> {
        return robotMongoRepo.findAll()
    }

    fun findOnlyAvailable(): List<Robot> {
        return robotMongoRepo.findOnlyAvailable()
    }

    fun updateAvailability(id: String, availability: Boolean) {
        mongoTemplate.update<Robot>()
            .matching(Query(Criteria.where(DocumentConstants.ID_FIELD).isEqualTo(id)))
            .apply(
                Update.update(
                    Robot.AVAILABILITY_FIELD,
                    availability
                )
            ).first()
    }

    fun addAbility(id: String, ability: Ability) {
        mongoTemplate.updateById<Robot>(id) {
            push(Robot.ABILITIES_FIELD, ability)
        }
    }
}